src/module.cpp

```diff
@@ -0,0 +1,677 @@
+{
+    std::vector<std::string> result = impl->input_names;
+    std::unordered_set<std::string> params;
+    for(auto&& ins : impl->instructions)
+    {
+        if(ins.name() == "@param")
+        {
+            auto&& name = any_cast<builtin::param>(ins.get_operator()).parameter;
+            params.insert(name);
+        }
+    }
+    erase_if(result, [&](auto&& name) { return params.count(name) == 0; });
+    return result;
+}
+
+instruction_ref module::get_parameter(std::string name) const
+{
+    auto ins = std::find_if(
+        impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
+            if(x.name() == "@param")
+            {
+                return any_cast<builtin::param>(x.get_operator()).parameter == name;
+            }
+            else
+            {
+                return false;
+            }
+        });
+    if(ins != this->end())
+        return ins;
+    else
+        return this->end();
+}
+
+std::unordered_map<std::string, shape> module::get_parameter_shapes() const
+{
+    std::unordered_map<std::string, shape> result;
+    for(auto&& ins : impl->instructions)
+    {
+        if(ins.name() == "@param")
+        {
+            auto&& name  = any_cast<builtin::param>(ins.get_operator()).parameter;
+            result[name] = ins.get_shape();
+        }
+    }
+    return result;
+}
+
+bool module::has_instruction(instruction_ref ins) const
+{
+    return std::find_if(
+               impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
+                   return std::addressof(*ins) == std::addressof(x);
+               }) != impl->instructions.end();
+}
+
+std::size_t module::size() const { return impl->instructions.size(); }
+instruction_ref module::begin() const { return impl->instructions.begin(); }
+instruction_ref module::end() const { return impl->instructions.end(); }
+
+std::vector<shape> module::get_output_shapes() const
+{
+    auto last_ins = impl->instructions.back();
+    if(last_ins.name() == "@return")
+    {
+        const auto& output_ins = last_ins.inputs();
+        std::vector<shape> output_shapes;
+        std::transform(output_ins.begin(),
+                       output_ins.end(),
+                       std::back_inserter(output_shapes),
+                       [](auto& ins) { return ins->get_shape(); });
+
+        return output_shapes;
+    }
+    // The else branch is to provide backward compatibility
+    else
+    {
+        return {last_ins.get_shape()};
+    }
+}
+
+instruction_ref module::validate() const
+{
+    return std::find_if(impl->instructions.begin(),
+                        impl->instructions.end(),
+                        [&](const instruction& i) { return !i.valid(impl->instructions.begin()); });
+}
+
+void module::finalize(context& ctx)
+{
+    for(auto ins : iterator_for(*this))
+    {
+        ins->finalize(ctx);
+    }
+}
+
+const int module_file_version = 1;
```

Comment:
The version should only be in the program, as this is used as a global-wide version to check for incompatibilities.